feat: 캐러셀 컴포넌트를 추가한다

chiabi opened the pull request.

chiabi committed:
feat: 캐러셀 컴포넌트를 추가한다

sim-zi commented on src/components/Carousel/index.tsx:
```diff
@@ -0,0 +1,278 @@
+}
+
+interface SwiperOnParameterType {
+  touchMove?: (event: any) => void;
```
event 타입은 알 수 없을까용?!

sim-zi commented on src/components/Carousel/index.tsx:
```diff
@@ -0,0 +1,278 @@
+  Light,
+}
+
+interface AutoplayParameterType {
```
인터페이스 이름 뒤에 Type은 생략해도 좋지 않을까욥!

sim-zi commented on package.json:
```diff
@@ -100,6 +100,7 @@
   "dependencies": {
     "lodash-es": "^4.17.11",
     "path-to-regexp": "^3.0.0",
+    "react-id-swiper": "1.6.9",
```
typescript으로 다시 짠 2.x 버전이 나온  것 같은데, 버전 올리는 게 어떤가유!

bae-unidev reviewed (approved):
조아!! 이거 더 develop해서 카로셀 안에 들어가는 카드 사이즈도 열거형으로 정해서 박아버리는게 어떨까?

chiabi committed:
feat: react-id-swiper 2.1.2로 업데이트한다

- update: react, react-dom >=16.8
- add swiper, @types/swiper

chiabi commented:
카드 사이즈 정하는 건 디자인에서 힘들지 않을까?
전에 lg: 4 column / sm: 2 column 이렇게 가는 것과 lg: 3 column / sm: 2 column 가는 것을 쓰임에 맞게 구분하는 이름을 지어서 정해버리기 어려울 것 같다고 얘기했거든

캐러셀 쓰이는 걸 보면 상품 캐러셀은 보통 4/2 인데 상세페이지에서 캐러셀이 여러 종류가 있고 perView를 auto로 주고 디자인대로 사이즈를 정해놓은 캐러셀들도 있어서(어떤걸 배우나요, 크리에이터 인터뷰 캐러셀)

chiabi committed:
refactor: Pagination, Autoplay 모듈을 추가한다

- getSwiper props을 추가한다

ddalpange reviewed (approved):
갓치치

sim-zi commented on src/GlobalStyle/swiperDefaultStyle.ts:
```diff
@@ -0,0 +1,313 @@
+import { css } from 'styled-components';
+
+/**
+ * Swiper 4.4.1
```
4.5.0버전으로 올리면 좋을 듯!

e.g. https://github.com/nolimits4web/swiper/blob/master/dist/css/swiper.css

sim-zi commented on src/components/Carousel/index.tsx:
```diff
@@ -0,0 +1,234 @@
+};
+
+export default class Carousel extends PureComponent<CarouselProps> {
+  public static ViewAllButton = ViewAllButton;
```
이거 나두 예전에 이렇게 썼었는데, 생각해보니 tree-shaking이 안 될 수도 있겠더라구... 흑
나중에 시간내서 class101-ui 컴포넌트들이 tree-shaking이 잘 되는지 점검해보면 좋을 듯!

sim-zi commented on src/components/Carousel/index.tsx:
```diff
@@ -0,0 +1,234 @@
+    );
+  }
+
+  private updateSwiper = (swiper: SwiperInstance) => {
```
updateSwiper라는 이름은 왠지 Swiper를 갱신하는(`swiper.update()`) 행동과 헷갈릴 것 같은데 좋은 이름이 없을까! (getSwiper가 명확하기는 할 것 같은데, getSwiper는 this.props.getSwiper랑 겹쳐서 못 쓴 거겠디..)


그런데 외부에서 getSwiper를 사용해서 swiperInstance를 쓸 일이 있을까?
웬만하면 swiper 자체는 숨겨놓고, 사용해야하는 메서드나 변수만 `<Carousel>` 에서 밖으로 인터페이스를 공개하는 게 좋을 것 같아

sim-zi replied:
지금 경우는 external library를 부모 컴포넌트에 전해주는 상황이라 예시가 좀 다르긴 하지만, ref에 대해서도 리액트 문서에서 웬만하면 캡슐화를 하라고 권장하구 있어!
외부에서 swiperjs를 직접 알지 않아도 되게 캡슐화를 잘 해두면, swiperjs에 의존하지 않아도 되기 때문에 나중에 다른 carousel 라이브러리로 교체할 일이 생길 때 고통이 없을 거야

> React components hide their implementation details, including their rendered output. Other components using FancyButton **usually will not need to** obtain a ref to the inner button DOM element. This is good because it prevents components from relying on each other’s DOM structure too much.

e.g. https://reactjs.org/docs/forwarding-refs.html#forwarding-refs-to-dom-components

chiabi replied:
흐으음 그럼 사용하게 될 메서드가 뭐가 될 지 아직은 잘 모르겠다. Infinite 캐러셀도 제거하고 전체보기 버튼으로 가자고 얘기가 되어서! 그럼 필요할 때 추가하기로 할게

sim-zi replied:
웅웅 getSwiper prop은 제거하구, 필요할 때 메서드 추가하잣!!

sim-zi replied:
아마 carousel.goNext() 나 carousel.currentSlideIndex 같은 게 필요해질듯..!

ddalpange commented:
이거 Lazy 옵션도 추가해줄수 있을까?

chiabi committed:
refactor: getSwiper를 제거한다

- ViewAllButton은 Button 컴포넌트를 사용한다
- getSwiper를 제거하고 필요한 변수와 메소드가 생길 때 추가하기로 한다

chiabi committed:
feat: lazy 옵션을 추가한다

ddalpange commented:
치치짱이야

chiabi committed:
feat: swiper 외부 네비게이션 버튼에 disbled상태를 추가한다
